### PR TITLE
Added softmax model

### PR DESCRIPTION
**Pull request by yangryan0**
Added a softmax model and integrated it with the PS. Uploaded the mnist dataset to S3, and the SoftmaxTask uses S3 for the dataset.

### CODE REVIEW

File: src/Dataset.cpp [hunk 1]
```diff
@@ -133,10 +133,36 @@ Dataset::build_s3_obj(uint64_t l, uint64_t r) {
...
-Dataset Dataset::Dataset::random_sample(uint64_t n_samples) const {
+Matrix Dataset::get_samples() const {
+  return samples_;
+}
+
+std::shared_ptr<const FEATURE_TYPE> Dataset::get_labels() const {
+  return labels_;
+}
+
+// SparseDataset Dataset::to_sparse() const {
```
Comment (jcarreira) [inline src/Dataset.cpp]: Clean these comments. There is more in other places.

File: src/LoadingSparseTaskS3.cpp [hunk 2]
```diff
@@ -26,16 +26,11 @@ SparseDataset LoadingSparseTaskS3::read_dataset(
...
-  return input.read_input_criteo_kaggle_sparse(
-      config.get_input_path(),
-      delimiter,
-      config);
+  return input.read_input_criteo_kaggle_sparse(config.get_input_path(),
+                                               delimiter, config);
 }
 
 void LoadingSparseTaskS3::check_label(FEATURE_TYPE label) {
-  if (label != 1.0 && label != 0.0) {
```
Comment (jcarreira) [inline src/LoadingSparseTaskS3.cpp]: I think we should check differently depending on the type of dataset. Removing these checks doesn't seems a good idea.

File: src/LoadingSparseTaskS3.cpp [hunk 3]
```diff
@@ -45,13 +40,12 @@ void LoadingSparseTaskS3::check_loading(const Configuration& config,
...
   std::cout << "[LOADER] Trying to get sample with id: " << 0 << std::endl;
 
-  std::string obj_id = std::to_string(hash_f(std::to_string(SAMPLE_BASE).c_str())) + "-CRITEO";
+  std::string obj_id =
+      std::to_string(hash_f(std::to_string(SAMPLE_BASE).c_str())) + "-CRITEO";
   std::string data =
       s3_client->s3_get_object_value(obj_id, config.get_s3_bucket());
 
   SparseDataset dataset(data.data(), true);
-  dataset.check();
```
Comment (jcarreira) [inline src/LoadingSparseTaskS3.cpp]: I think these checks are useful and should be kept unless there is a good reason to remove them

File: src/LoadingSparseTaskS3.cpp [hunk 4]
```diff
@@ -67,9 +61,6 @@ void LoadingSparseTaskS3::check_loading(const Configuration& config,
   for (uint64_t i = 0; i < dataset.num_samples(); ++i) {
     //const auto& s = dataset.get_row(i);
     const auto& label = dataset.labels_[i];
-    if (label != 0.0 && label != 1.0) {
```
Comment (jcarreira) [inline src/LoadingSparseTaskS3.cpp]: Same here. These checks are useful.

File: src/LoadingSparseTaskS3.cpp [hunk 5]
```diff
@@ -108,7 +99,9 @@ void LoadingSparseTaskS3::run(const Configuration& config) {
       dataset.build_serialized_s3_obj(first_sample, last_sample, &len);
 
     std::cout << "Putting object in S3 with size: " << len << std::endl;
-    std::string obj_id = std::to_string(hash_f(std::to_string(SAMPLE_BASE + i).c_str())) + "-CRITEO";
+    std::string obj_id =
+        std::to_string(hash_f(std::to_string(SAMPLE_BASE + i).c_str())) +
+        "-CRITEO";
```
Comment (jcarreira) [inline src/LoadingSparseTaskS3.cpp]: I think this "-CRITEO" should be removed altogether from the code.

File: src/PSSparseServerInterface.cpp [hunk 6]
```diff
@@ -74,6 +74,34 @@ void PSSparseServerInterface::send_lr_gradient(const LRSparseGradient& gradient)
   }
 }
 
+void PSSparseServerInterface::send_sm_gradient(
+    const SoftmaxGradient& gradient) {
+  uint32_t operation = SEND_SM_GRADIENT;
+#ifdef DEBUG
+  std::cout << "Sending gradient" << std::endl;
+#endif
+  int ret = send(sock, &operation, sizeof(uint32_t), 0);
```
Comment (jcarreira) [inline src/PSSparseServerInterface.cpp]: We should always use send_all.

File: src/PSSparseServerInterface.cpp [hunk 7]
```diff
@@ -197,6 +225,31 @@ std::unique_ptr<CirrusModel> PSSparseServerInterface::get_full_model(
...
+
+  char* buffer = new char[to_receive_size];
+  read_all(sock, buffer, to_receive_size);
+  // std::cout
+  //<< " buffer checksum: " << crc32(buffer, to_receive_size)
+  //<< std::endl;
+
+  // build a sparse model and return
+  std::unique_ptr<SoftmaxModel> model = std::make_unique<SoftmaxModel>(
+      (FEATURE_TYPE*) buffer, config.get_num_classes(),
```
Comment (jcarreira) [inline src/PSSparseServerInterface.cpp]: I would make buffer a std::shared_ptr<char[]> and use a dynamic cast here.

File: src/S3SparseIterator.cpp [hunk 8]
```diff
@@ -235,6 +234,9 @@ void S3SparseIterator::thread_function(const Configuration& config) {
     if (config.get_s3_bucket() == "cirrus-criteo-kaggle-19b-random" ||
       config.get_s3_bucket() == "cirrus-criteo-kaggle-20b-random") {
       obj_id_str = std::to_string(hash_f(std::to_string(obj_id).c_str())) + "-CRITEO";
+    } else if (config.get_s3_bucket() == "cirrus-mnist") {
```
Comment (jcarreira) [inline src/S3SparseIterator.cpp]: This might be a good time to fix this old code I badly wrote.

File: .gitattributes [hunk 9]
```diff
@@ -0,0 +1 @@
+*.csv filter=lfs diff=lfs merge=lfs -text
```
Comment (jcarreira) [inline .gitattributes]: Didn't we talk about not using git lfs?

File: python/frontend/cirrus/examples/SoftmaxDemo.ipynb [hunk 10]
```diff
@@ -0,0 +1,1073 @@
...
+     "output_type": "stream",
+     "text": [
+      "scp machine_0.sh ubuntu@ec2-18-237-213-139.us-west-2.compute.amazonaws.com:~/\n",
+      "scp machine_1.sh ubuntu@ec2-18-237-31-107.us-west-2.compute.amazonaws.com:~/\n",
+      " * Serving Flask app \"cirrus.graph\" (lazy loading)\n",
+      " * Environment: production\n",
+      "   WARNING: Do not use the development server in a production environment.\n",
+      "   Use a production WSGI server instead.\n",
+      " * Debug mode: off\n",
+      "('Thread', 8, 'exp', 8, 'loss', [])\n",
```
Comment (jcarreira) [inline python/frontend/cirrus/examples/SoftmaxDemo.ipynb]: The notebooks should be only code.

File: src/Dataset.cpp [hunk 11]
```diff
@@ -76,10 +76,11 @@ uint64_t Dataset::num_samples() const {
     return samples_.rows;
 }
 
-void Dataset::check() const {
+void Dataset::check(const Configuration& config) const {
   const FEATURE_TYPE* l = labels_.get();
   for (uint64_t i = 0; i < num_samples(); ++i) {
-    if (!FLOAT_EQ(l[i], 1.0) && !FLOAT_EQ(l[i], 0.0)) {
+    if (!FLOAT_EQ(l[i], 1.0) && !FLOAT_EQ(l[i], 0.0) &&
```
Comment (jcarreira) [inline src/Dataset.cpp]: Shouldn't we also check labels when we use softmax?

File: src/DatasetConversion.h [hunk 12]
```diff
@@ -0,0 +1,12 @@
+#ifndef _DC_H
```
Comment (jcarreira) [inline src/DatasetConversion.h]: DC is not a good name

File: tests/test_travis/mnist_test/.deps/sm_test.Po [hunk 13]
```diff
@@ -0,0 +1,1484 @@
+mnist_test/sm_test.o: mnist_test/sm_test.cpp /usr/include/stdc-predef.h \
```
Comment (jcarreira) [inline tests/test_travis/mnist_test/.deps/sm_test.Po]: What is this?

File: configs/softmax_config.cfg [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+load_input_type: csv # for the criteo kaggle train.csv
+num_classes: 10
+num_features: 784
+limit_cols: 1000
+normalize: 1
+limit_samples: 50000000
+s3_size: 1000
+momentum_beta: 0.9
+use_bias: 1
+opt_method: momentum
```
Comment (jcarreira) [inline configs/softmax_config.cfg]: If there arguments here that we don't use for softmax we should remove them. Also, sgd should be the default optimization method.

File: python/frontend/cirrus/cirrus/core.py [hunk 15]
```diff
@@ -126,8 +126,8 @@ def get_updates_per_second(self, fetch=True):
 
     def relaunch_lambdas(self):
         if self.is_dead():
+            print "dead"
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/core.py]: This line doesn't make sense.

File: python/frontend/cirrus/cirrus/sm.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+from core import BaseTask
+
+
+class SoftmaxTask(BaseTask):
+    def __init__(self, *args, **kwargs):
+        # pass all arguments of init to parent class
+        super(SoftmaxTask, self).__init__(*args, **kwargs)
+
+    def __del__(self):
+        print("Softmax Task Lost")
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/sm.py]: This print is not very informative.

File: src/Dataset.h [hunk 17]
```diff
@@ -118,7 +121,13 @@ class Dataset {
    */
   Dataset random_sample(uint64_t n_samples) const;
 
-  public:
+  Matrix get_samples() const;
+
+  std::shared_ptr<const FEATURE_TYPE> get_labels() const;
+
+  // SparseDataset to_sparse() const;
```
Comment (jcarreira) [inline src/Dataset.h]: This line should not be here

File: src/DatasetConversion.h [hunk 18]
```diff
@@ -0,0 +1,12 @@
+#ifndef _DATASETCONVERSION_H
+#define _DATASETCONVERSION_H
+
+#include <Dataset.h>
+#include <SparseDataset.h>
+
+namespace cirrus {
+SparseDataset to_sparse(const Dataset& d);
```
Comment (jcarreira) [inline src/DatasetConversion.h]: Please add documentation to these functions.

File: src/ErrorSparseTask.cpp [hunk 19]
```diff
@@ -54,9 +55,12 @@ std::unique_ptr<CirrusModel> get_model(const Configuration& config,
       }
     }
   }
-
+  bool use_softmax = config.get_model_type() == Configuration::SOFTMAX;
   bool use_col_filtering =
     config.get_model_type() == Configuration::COLLABORATIVE_FILTERING;
+  if (use_softmax) {
+    return psi->get_sm_full_model(config);
+  }
```
Comment (jcarreira) [inline src/ErrorSparseTask.cpp]: An else after the if makes this more readable.

File: src/ErrorSparseTask.cpp [hunk 20]
```diff
@@ -107,7 +111,8 @@ void ErrorSparseTask::error_response() {
   }
 }
 
-void ErrorSparseTask::run(const Configuration& config) {
+void ErrorSparseTask::run(const Configuration& config, bool testing) {
+  s3_initialize_aws();
```
Comment (jcarreira) [inline src/ErrorSparseTask.cpp]: I suggest we put an s3_initialize_aws() before run_tasks in parameter_server.cpp and remove this.

File: src/S3SparseIterator.cpp [hunk 21]
```diff
@@ -154,7 +154,6 @@ void S3SparseIterator::pushSamples(std::ostringstream* oss) {
     for (uint64_t j = 0; j < minibatch_rows; ++j) {
       if (use_label) {
         FEATURE_TYPE label = load_value<FEATURE_TYPE>(s3_data); // read label
-        assert(label == 0.0 || label == 1.0);
```
Comment (jcarreira) [inline src/S3SparseIterator.cpp]: I would like to leave some code here to check the labels (inside a #debug pragma).

File: src/SparseDataset.h [hunk 22]
```diff
@@ -102,6 +105,8 @@ class SparseDataset {
   
   SparseDataset sample_from(uint64_t start, uint64_t n_samples) const;
 
+  Dataset to_dataset(const Configuration& config) const;
```
Comment (jcarreira) [inline src/SparseDataset.h]: This should have a comment.

File: tests/test_travis/mnist_test/sm_worker.cpp [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+
+  std::unique_ptr<PSSparseServerInterface> psi =
+      std::make_unique<PSSparseServerInterface>("127.0.0.1", 1337);
+  psi->connect();
+  int version = 0;
+  while (1) {
+    Dataset minibatch = train_dataset.random_sample(20);
+    SoftmaxModel model = *(psi->get_sm_full_model(config));
+    auto gradient = model.minibatch_grad(minibatch.get_samples(),
+                                         (float*) minibatch.get_labels().get(),
```
Comment (jcarreira) [inline tests/test_travis/mnist_test/sm_worker.cpp]: Use c++ cast.

**Comment by jcarreira** [comment]
Please fix conflicts.
